### ISSUE TITLE
The KDE transform creates values where there are none when used with `{"resolve": "shared"}`

### ISSUE THREAD

**Issue by joelostblom** [vega/vega#3815]
If  `{"resolve": "shared"}` is set, the extent of grouped density transforms incorrectly use the min/max of the entire dataset instead of for each group, resulting in long lines where there are no observations at all, instead of stopping the density at the last data point in the group. I noticed this in Vega-Lite, but wonder if it could be fixed directly in the KDE transform in Vega instead of doing some post-processing such as dropping zeros in Vega-Lite. I understand that the computation need to happen over the same domain to enable stacking, but would it be possible to trim the densities after that to only include values that exists within each group?

This chart is created in altair 5.1.2 which uses VL 5.15.1 and shows the undesired behavior:

![image](https://github.com/vega/vega-lite/assets/4560057/470d6e00-9a60-4b58-95c4-ff1ec5dc9f2f)
[Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4Igxg9gdgZglgcxALlANzgUwO4tJKAFzigFcJSBnAdTgBNCALFAZgAY2AacaYsiygAlMiRoVYcAvpO50AhoTl4QpAE4AbFCDGEADpWQB6Q2DpQAdACtKdTOrhpV5qJkKGougLaG0mBHIBaeUVKV0oAATQARnMAJgBOczZDYLlDOFU4SitKaBAZEE85VQBrZUIAT11MLWLMJW5KQlUIEprkEAAjdTkwMoLMKEg6EiRUEBhe12VIdVJPKEoUKO54OzotSmqwLCXuRnrbVWUezrsAMV5qEQQxLU6IdQ3uU4veAGU4AC92qNiC4iEdTtMjqdTcSrVLRQCCeEhyTQFAAeyjWTy0aARpBqEKq7RAAEdSHIiHBFMRfPluBVUVh0R1bIsyTTGoo+ihJupQrioR0iSTAQoHDVpPsbndkFEOBDVCTKDAIKpPCgANqgRmUZlaaqKdS0BjMbgIFqkXSdGnIFUgLaYHaYJYAXUkDu4ABJKGADkUtDp9EYfH45OYEGTGKROuY4BBDB6vWlfP4AvZCJgfABWcxRDMxax5SRAA) 

The desired behavior would look like this where each density is cut at the min/max values of each group:

![image](https://github.com/vega/vega-lite/assets/4560057/3f759c0e-a610-46b6-87a4-e676e2ad2e48)

<details><summary>Altair code</summary>

```py
import altair as alt
from vega_datasets import data

source = data.iris.url

alt.Chart(source, height=100).transform_density(
    'petalWidth',
    groupby=['species']
).mark_area(stroke='black').encode(
    alt.X('value:Q'),
    alt.Y('density:Q').stack(False),
    alt.Facet('species:N', columns=1, title=None).header(labelFontWeight='bold', labelFontSize=12)
)
```

</details>

Ref https://github.com/vega/vega-lite/issues/9078

**Comment by joelostblom** [comment]
@jheer Do you think this is something that is suitable for implementation on the Vega side of things or does it belong in the Vega-Lite repo?

A related issue stemming from this is that setting the x-scaled to `"independent"` does not have the intended effect. Take for example this chart where I would like the axis to be adjusted in each subplot to only span the range of the data, so that I can see both distributions clearly:

![image](https://github.com/vega/vega/assets/4560057/303a1962-19c7-4c54-b3aa-aa2a757d03a4)
[Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4Igxg9gdgZglgcxALlANzgUwO4tJKAFzigFcJSBnAdTgBNCALFAZgAY2AacaYsiygAlMiRoVYcAvpO50AhoTl4QpAE4AbFCDGEADpWQB6Q2DpQAdACtKdTOrhpV5qJkKGougLaG0mBHIBaeUVKV0oAATQARnMAJgBOczZDYLkTOVVKK0poEBkQTwyAa2VCAE9dTC0MzCV8zChIOhIkVB51Uk8oZXg7Oi0izDKQbnLKrSgITxI5TXyADx6sdX7kEDRZ0irRiqq1gEdSOSI4RWJfPO5htt6VrVsoSlPhnfGDo5OzhyrpblVMHLqC5tShgWZ7UCLNYkWyVKAPcS-ECEVTHSgwCCqTwoADaoAxd2QOJAAEEwGA7JhUcRctxqCIEGIAPokJnqABGlBAAF0ZKAHk9yloNh1tiAEKoKLp2ddiYNhrzudwACSgxiYQpaHT6Iw+PxycwIU6MUjs8xwCCGNUatK+fwBeyETA+ACs5iibpi1lykiAA)

**Comment by mattijn** [comment]
If I add a `"y":"independent"` to the scale-resolver in the VL-spec and in the Vega-spec remove the `impute` transform and make the `kde` transform to resolve `independent` it seems like what you are after:
<img width="434" alt="image" src="https://github.com/vega/vega/assets/5186265/57e2b157-7519-4679-bd26-7d288641493b">

[Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhBMkUANZkATmwbiAJmhAB3GHTjSQOJBo01xZNAJkakmFKgDaocUgSH0YhsqhwA+gAMRj4ANtpYuPhEUBriIhAacKE0xMqU4nCYBOI4CLwUALR2DhBZEAACxACMlABMAJyUgQQl-FBIyhAJEkYAZmzKyJhooJgAnjieIKK9AL4ymMpI4hADQ2guIBNT2gOhWjJ9NMkaEJsgAIJQfqFwy5jskjIA6nA0ZFj+Fv6hTOcAXRkSHOzhACjg4yMxCQoQYhgBCzGk2mCiS-ROB20MLhhhkKjUOCYULBEKhQJAIIuOPhRiSqzo5JkyjgYlCpG0FiSU00cHEIyR2xR2ggDkURgJDCJJKcIBpePBkJAFOOp209IgjKMYmUI1QoFVWOcFMGSWUm0RwNBso1jP8os6I1sfM1E38fK0Jr6fTKepAAC97mwQHNLa53NMoGw4QhxP4NGxkBZtWpfNNvGmgkYlis1oMEJtkbt0EgyCo+JgFZLpRcycrQ6GZJqyG5Qtbwx5tLAaAd-DoaBp6NDYbTUNUBIFBW5O+hu734B8sMPcWhx5OKaEkOM1HrQCYzBYrKg6oEZGxvb7RiAozHxAAVOh3VeTmQ3hix0GgZut7R3Sz0AAKNoAIAcjfWN40TJALBAgBKWCQxkJg1E0UEQD6BhQnCYEUhbbRxAkQwFhAZBlAUdsQGnSNo3fcRCkeTAn0WYV0GrIxVCfWcaNjejHwVBjOLGOAAA8-TrJsJk4kAyAYAc4F4xiFXPH0smfOYpwjLtuLjeBTHubMWOk1QpXY6NqNvQpdLNfpVALfUQDaLTb0gpNJGInVd3QzEtHQEp3ycAAdRVxiCgEjFNfSS2gD1D0Q7Y+KvStRKvb9YU5CAADVYQHACTBZflAuC0LYIAAgAfhKvK+UwQqyVCkrUBKoKgoAaiqgqgrq5U4tFcYpJkuTCk3JhkhszTWOMnBwovVSxxfEA+SjdF7KlEpPFAftBxgFKPh-Wd9F7Tah3U9SpA7SNkmwoViyMwltUki6sJsxMrz6eRZvOvZ3qdBz7EcEAMz8LN8Um4layVRF3MGTzDR8m17H8zrIVC5Uz2UaywRBPxNFiy0FvEJb1pAVb7CJo7tvs1Lwn2nsND7AdjpkBdPk8qmuwOunmaXE6ZFI8jC0o8brw5-w+fOZibs6OAUAkvrPFlKWUApDzXu89UEYQQr5RR4i+lsq9HPQN6-AFGRFrYZbQBJysr0GE5+SveVsXuR4OnCXWe2p0AnfQABiAAWKAAHYAA4kBDuKkmgZQaBwR5ekp3a0vQIL5UaoKSpakqAPWYYgI1rWRzgUKpCakAgvgzOy4AbhK20JnTkAq5z-N7HzzBEYcl1GRLsuK4Q4jhJSt3IxFoejjV9AneIkkvxH9naf8KEJ7VXzu7dD04vGOph9hUfF+XrzV67hk3QdXVI7gY5Mh8r8k+pkAaEy7KNHbzvteVUqADIv5Kp+ADELAGAAi1PymtU5F2KiVH+f9n4pFfmAwq9cQqf2gb-ABQDKwgMQUFZBxUQwnWViPCiVEF69nHtdaYKRMidDpFBZM9lDboW+hiY+09mQrDIPLU8d8WzJ2FovcmIYKTiBoH4NASx4QyEDKoNAb02xwA0jOARvZD47CoRYaW5pbD0MkIwv6X0TasIONaE+rpxj2gcBfZ0p8LGbzxssSw8teF7RUZzd4LNEKBBEWIzwki4DSKDBI5Q8JGyUmEqyAW0A95kLphQu21VtDIWwC9EGA4ADK88aaqIlDHHy-jFhiIUAAYRQqze+XZ3ihAAnOOm5MCD+0CAPHRrk5GwjKDIYayRQTyI6ZSGOjhekBJIkgYSABRUSiTUCnhIhYCZlYHbTMKYoHp7Thn+i5CJNAL454xOyXTQ+CTFkgDuH0H6KgMlZLcf4ChFz8khOGa7EpZSdp8Ifn4HsNSRZc2yI05pDldFtIUZ0pAI02xAr6Z0Gggy1m81GfMqZMyEBzMmYsmZTzVnAoDJsoe0zBTRKkrUm54UY5TNkGwFJBY0k+SGYUxS2Ii5GC6aEf+cIIAUwKSc0FyQADypBlCbimqgTlTzSnqHKW8ypnyiX1L+XFDZvJcU7IBlcolhzSXHNOecvJELHkJTXrYpl3LQh8vuIK4JUj4qKDFYslx-CPnVKJT8hpTT5U4u2Q2REQA)

In my opinion this is something for the the VL-repository.

**Comment by joelostblom** [comment]
Thanks @mattijn, setting the transform resolve to independent would fix both the specs above, but it would lead to jagged appearance when having two densities in the same chart as described in this issue https://github.com/vega/vega-lite/issues/9078. So we would either need another way to fix that (maybe setting the steps + the extent?) so that we can use `"resolve": "independent"`, or make the shared resolve more flexible so that it works with the examples in this issue. I'm happy with whichever solution is the easiest to implement and support these use cases.

**Comment by joelostblom** [comment]
After investigating this further, I can give a more comprehensive explanation of what is going on. Here is a single spec that contains both issue. I can't find any combination of parameters that supports each density ending at the min/max value of the data AND being able to have the two grouped/colored densities display properly on top of each other.

Step 1: Coloring by one variable and faceting by another. You can see how the lower facet ("Open") is extended all the way to the x-axis min around 3.5, although there are not data points there. (Also note that by default Vega-Lite now stacks areas which is not ideal for distribution densities since it makes them harder to compare, but this is a separate issue https://github.com/vega/vega-lite/issues/9170).

![image](https://github.com/vega/vega/assets/4560057/ef86cb9f-ef61-4884-822a-2afd8b4ea7ce)
[Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega-lite/N4Igxg9gdgZglgcxALlANzgUwO4tJKAFzigFcJSBnAdTgBNCALFAZgAY2AacaYsiygAlMiRoVYcAvpO50AhoTl4QpAE4AbFCDGEADpWQB6Q2DpQAdACtKdTOrhpV5qJkKGougLaG0mBHIBaeUVKV0oAATQARnMAJgBOczZDYLlDTwgMTEorSmgQGRAYOTBXZXg7Oi0AWQAFAEF6gH0AJQUSJG5CAE9dTC0oCE8SOU1Cyj6wZU85VQBrZR6+rVnMJW4IXRK4HpQ2cwAObkpCVQg5-uQQACN1EoXj0-PMWgZmZFjCzChIOg7lSDqCCqcpYdRVK7VOSWYFNADi31U-S6vUuIEGwygowK3AAHqDKlo0KNSMiQEs0QBHUhyIg7dq+HEgboE8FaWxQSg7Fko5ZXam04iKYiM6TcRgiBBiFBRDhdVS0ygwYGeFAAbVA8HUhEwINQIFpEI1RTBEJAUJhqnhiLJ0EwAHkYOqQPUwMR8twACIKmYgAC6Mk1ppqDWabT4nRAdsdzvtfSgIG4ADkIIQAAThzBVAMBwMgDlc3ZXACS1U9ACFWu0oJGEGdSLprizkGrzdDYQioEjE+bQ1WI-7JAHuIDSJ5OTLuAASShgCW+q46fRGHx+OTmBA7Rika7mOAQQyz+dpXz+AL2HU+ACs5iiN5i1nykiAA)

Step 2: I can fix the issue with the extension to zero if I set the resolve to independent and remove the impute transform as you suggested. However, that automatically unstacks the areas (which often is a good default but would be unexpected to someone who explicitly specified a stacked density:

![image](https://github.com/vega/vega/assets/4560057/9d5a4e0d-f44b-4961-b5af-760b4e898a6e)
[Open the Chart in the Vega Editor](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhBMkUANZkATmwbiAJmhAB3GHTjSQOJBo01xZNAJkakmFKgDaocUgSH0YhsqhwA+gAMRj4ANtpYuPhEUBriIhAacKE0xMqU4nCYBOI4CLwUALR2DhBZEAACxACMlABMAJyUgQQl-AhsxDRwEAkSRgBmbMrImGigmACeOJ4gov0yJspl4yAAskjCw-4A4nDiyrMQmMoWVjJrAAoAgtf+AEr2Z9rHp5YgAL4fMidI4hBDEZoFwgKYzbQDGihTBwZRGOAADxwcPQAAoLEkEWwBqinAAdEDXKCYdjiAlSAkAEWU7iQBIAulIAAQlBgIfHrTbbPYHOAMgCUTIAhABeEVMwrVQUAMmlTPRmkR2NxBIA8jMydICQA5NiYJmPGEaBnM1nsglXW4PJ6WAXCsUSqWfKQTaazBRJQbdUJadAASTWlIAQtaSe8ZCo1DgmJNgZytspdvtDkZLXdDc9GSAkBA48QkKEGIZbPsIHRY1nDmJQqRtBi4BqkuIxt9XeCvA5FEZIwxo7HnCB84Xi+sbumbVYs5Dkr6QE2y1MjGJlGNUKBpz64xsE0neSAs8MkiinHOen5NJnvtncwP5+X-MckCujHepv59loDwMBmVVyAAF6wmwnz0q2IBuB4ELyFk-gaGwyAWEuai+EcyF+EERi-P8gIIMCbazEgZAqHwMLdqovYxluY6hpeoEuuB7izAM0GYLB8FIBY-iqDomE0thwy4c4+EvHAACORbiH4S4OM+qCBDIxxsDgqxlmQbjhOgfhQqioT7GQ9Com0qIEsxfisXBCGavygoEEyTpfHRriMVBZlsZZ-hQGwhYIJIPx8QCAl4aCboieJ+xSQpMmrvJICKcpa6xTQakFtoCAWDpekGUZJksW5HFWfyzL2R8oFZqp6k3iCEGzLAUIaP4Og0Bo9BGEORZoAAzIEgRgdV2i1T6-jwElWCtQW7WoNU3UlTIoRIJMairqAJhmM8qB1DFTBqJoN4gAMDChOEMgFklkjoAWR0gJ53n-Gg1RXsgygKJVTmQegplZIUJKYLpmEhegPbKTIqi-ZpXlsuIX10L9PzQ54EyIn+aY0eGsVTKDIBkAwTVwFDP0jtiP5ZHdPW9c56DcUNcCmLCf3tpj5FAyAIOzNxhTwDTcIyAMqiCaAbQuTBFn5VxbA8Ve4VsJ6CW9iU8MYHAI1LYlyUaVd+iDcNZCjV8V5IAiPRBdABazLGMjDN0zbaLpAxjBGpy+sxoRlLD+PaK+ZsgHNTDJKqpDKHN8UnEWsOKAAwttyvlSlmmK6EqIDfVWtYAQAAsgT8s6AH1giaA9Y5DFvVd4M+f4Qx6rTPz-QzUZGCz-Ul5D5cwlze286sAvvblwucddENZ5L0ugLL9jy41zUwCpp0x+rdUNU1LW68dBsvbFUAm9oufm6c+x-lt2DwWRTVoE7LugnD2htSO3vJAAYoWECT6gwdwLNSA+6EfuwoHaAv6HCgR3UFHaeastLx0TvPCeacM5Z3-DnPOJUyZFz8IdOmsxAaYQvglGECJgGqzrBAAAaidDQqIli7w5MjDMtp9yCgAPxMnIc2Sh1FqFkAZEyVATICQEgANRMMwCwq0bCGRZ2OJMDGWMcaFBvpdHm5Ma69iMITX8JMrziIxiguR7cEofWVn1LuZkXz2EcLFNCARgj2yjJRAcVCJz7i+ApYYysNy+hPGaIR44wzsP3PubeR44w5nPOYd4dEQCD3liPUiCVx4tQStHUBGt6qxMnleZOeD1L9SSVTJWnxHEgEes9IKBjZ6DRMPQDBVd6aVLbofXRLFVglL0XAeq5SYAeWyZZYxDhBZ20UX2LcXJEw8hTKBCWkkpaRJwHLVYKTViuNWBg9Acz8npPmd6X0oAlkK1ybrB6T4ilCULjVTpBzcxVIIocFAkUJGeBPE+amfizGyXXBs929g2QcivqIq8PM6n8xMb0lp-g2kdLnl08Znkh4gCifLJS8hyyrCvnnSgAAOdRJw2AKHlsi9ATA5pdgxaobFAB1BeT9QC4rqFeC2u8kXjVmP7Ek69wi-KhBpUAxtNFeWGF6Gc2htzcmTIYK8SRoCnBwCSfo8SQHaAJFfLhBImS8PlDhewhkPnmkHAyk03CQAEmssqvVABuFkpZyyKpAEa1EarMAaswJ8gkHtdU8JAIalVBJTWCuGcKy11qaBEJIfax18YhW8jtAwjxFohm7kOBwy1fCo2hp9eG2hWdc7xJZbMLee03noGRVefsnKs3aE9gs9AHt3yaCzpMOoKkS3oDLXm08-x7yPmfKKuAkJMibJVpk9AAbiEpFIUm75tCmSyiZAG2+FgDCol4aOnV47J2DqDUm51y65TTtnTCed67zVTAFHkhyM1swryNg2kAObaVWzxXqTAh97ZNQAMqXs9ioY+z9lAh3PuHSOU98Gx20hAlJ0DM6ivYohVAp9X5e3fskG8MHjqnEcEhgp+sACiuC6VyRkGlcQWGYS3pisyw5aG4GKgzT1eiXLTbKJ3rer2Xa+kfo0K+je6Ac2sd-t+2DpHAG3s5bKoD4DsnpLA1nHuZ00OyMQwWM+T4aCofk7B5ACJCM4ZivhjTxH-5yedrBijmIEGlQUlm1eJSc1glmCkTIT4XyQbOgCnpXhzEYW5s2gtwM-hkDufJITgHSnJPJSBGQ4gaBSS-T+wCqgT4qaQXRi51sLDU1blJjugLXM+HQpY3NM5KotoXJMB8UUXwHuKx+Bx3nLB3IC-2oLOTtYtikIELM4XIt-wAkBHjRYEsN1CLypL6BDwWBSrYRzGWXNmOyxYvlm50DetjSOZcq4X5Xj4r5-qo8yDDFjKe3SvmdpBUhKgsGA3W4QEmAgLaoQAAq1coA0F8DDc+bsFsxpGSOCJKkrs3ZvMPaZo8VKYuxfS4c2h8XyAUFneFj3FwJVxYENFezEFAA)

In other words, I can't find a combination of parameters that allows me to create this chart (correctly stacked on top, and correct extent on the bottom):

![image](https://github.com/vega/vega/assets/4560057/446c2c8e-0609-4b68-aeb7-59275579cbdf)

**Comment by domoritz** [comment]
It looks like the first case can [be resolved](https://github.com/vega/vega/issues/3815#issuecomment-1805280690) with explicitly setting the kde resolve which we add in https://github.com/vega/vega-lite/pull/9172

https://github.com/vega/vega/issues/3815#issuecomment-1806904528 is a bit trickier but could be addressed with a `clip` property that removes density values outside the original data domain per group. This could be a useful feature anyway (for both shared and independent density computation.

**Comment by domoritz** [comment]
I'm working on this now
